### PR TITLE
allow configuring a target chunk size in compressed bytes

### PR DESCRIPTION
**Pull request by slim-bean**
Need to figure out still what to do with the error if you fail to cut a chunk...

Signed-off-by: Edward Welch <edward.welch@grafana.com>

### CODE REVIEW

File: pkg/chunkenc/gzip.go [hunk 1]
```diff
@@ -329,11 +348,25 @@ func (c *MemChunk) UncompressedSize() int {
...
+	}
+	size += c.cutBlockSize
+	return size
+}
 
-	return float64(size) / float64(blocksPerChunk*c.blockSize)
+// Utilization implements Chunk.
+func (c *MemChunk) Utilization() float64 {
+	if c.targetSize != 0 {
+		return float64(c.targetSize) / float64(c.CompressedSize())
```
Comment (rfratto) [inline pkg/chunkenc/gzip.go]: I'm not sure this is right, shouldn't it be `compressed_size / target_size`?

File: pkg/ingester/stream.go [hunk 2]
```diff
@@ -106,13 +116,18 @@ func (s *stream) Push(_ context.Context, entries []logproto.Entry) error {
 	for i := range entries {
 		chunk := &s.chunks[len(s.chunks)-1]
 		if chunk.closed || !chunk.chunk.SpaceFor(&entries[i]) {
+			err := chunk.chunk.Close()
```
Comment (rfratto) [inline pkg/ingester/stream.go]: I'm not clear why we need to cut the head block now but we didn't have to before, could you explain?

File: docs/configuration/README.md [hunk 3]
```diff
@@ -267,8 +267,16 @@ The `ingester_config` block configures Ingesters.
 # period as long as they receieve no further activity.
 [chunk_idle_period: <duration> | default = 30m]
 
-# The maximum size in bytes a chunk can be before it should be flushed.
+# The maximum _uncompressed_ size in bytes of a chunk block,
```
Comment (rfratto) [inline docs/configuration/README.md]: I don't think this is 100% accurate: almost all blocks are going to end up being greater than this size. This is more like the cuttoff point for when an uncompressed block should be compressed. 

Do you think it's worth explaining that in the docs?

File: pkg/ingester/stream.go [hunk 4]
```diff
@@ -106,13 +116,18 @@ func (s *stream) Push(_ context.Context, entries []logproto.Entry) error {
...
+			if err != nil {
+				//TODO what the heck do we do with this error?
```
Comment (cyriltovena) [inline pkg/ingester/stream.go]: log it for sure.

File: pkg/chunkenc/gzip.go [hunk 5]
```diff
@@ -309,9 +315,22 @@ func (c *MemChunk) Size() int {
...
 // SpaceFor implements Chunk.
-func (c *MemChunk) SpaceFor(*logproto.Entry) bool {
-	return len(c.blocks) < blocksPerChunk
+func (c *MemChunk) SpaceFor(e *logproto.Entry) bool {
+	if c.targetSize > 0 {
+		// This is looking to see if the uncompressed lines will fit which is not
+		// a great check, but it will guarantee we are always under the target size
+		newHBSize := c.head.size + len(e.Line)
+		return (c.cutBlockSize + newHBSize) < c.targetSize
+	} else {
```
Comment (cyriltovena) [inline pkg/chunkenc/gzip.go]: the else is not necessary

File: pkg/chunkenc/gzip.go (hunk 1 above)
Comment (cyriltovena) [reply pkg/chunkenc/gzip.go]: and that else though.

File: pkg/ingester/stream.go (hunk 2 above)
Comment (slim-bean) [reply pkg/ingester/stream.go]: by the nature of how we calculated the SpaceFor previously, it was always guaranteed to have cut the head chunk:
  * append a log line, if that log line is > than the block size, cut the block.
  * next log line, check to see if the number of blocks is >= max blocks per chunk, if yes, make a new chunk.

When we are checking SpaceFor against a target size there head block could still have entries in it but the new line could push us over the edge, so we have to make sure to cut what's out of the head block.

I debated if the SpaceFor could incorporate this behavior but that felt like kind of an ugly side-effect of a method which I wouldn't expect to make any changes to the chunk.

File: docs/configuration/README.md (hunk 3 above)
Comment (slim-bean) [reply docs/configuration/README.md]: yeah agree, will reword this

File: pkg/chunkenc/gzip.go (hunk 5 above)
Comment (slim-bean) [reply pkg/chunkenc/gzip.go]: yeah, our linter agrees with you but personally I prefer it this way... but I guess it's me vs you and the linter and I'm gonna lose this one :)

Comment (cyriltovena) [reply pkg/chunkenc/gzip.go]: You are !

File: pkg/chunkenc/gzip.go (hunk 1 above)
Comment (slim-bean) [reply pkg/chunkenc/gzip.go]: compresses size = 500kb
target size = 1000kb

Current Method Utilization = 1000/500 = 2... You win this round @rfratto

File: docs/configuration/README.md (hunk 3 above)
Comment (slim-bean) [reply docs/configuration/README.md]: @rfratto reworded, let me know what you think

File: pkg/ingester/stream.go (hunk 2 above)
Comment (cyriltovena) [reply pkg/ingester/stream.go]: good for me.

Comment (rfratto) [reply pkg/ingester/stream.go]: :+1:

File: docs/configuration/README.md (hunk 3 above)
Comment (rfratto) [reply docs/configuration/README.md]: Nice, this is more clear. Thanks!

**Pull request merged**